SI-8134 SI-5954 Fix companions in package object under separate comp.

retronym opened the pull request:
Review by @odersky

retronym committed:
SI-8134 SI-5954 Fix companions in package object under separate comp.

The tests cases enclosed exhibited two failures modes under
separate compilation.

  1. When a synthetic companion object for a case- or implicit-class
     defined in a package object is called for,
     `Namer#ensureCompanionObject` is used to check for an explicitly
     defined companion before decided to create a synthetic one.
     This lookup of an existing companion symbol by `companionObjectOf`
     would locate a symbol backed by a class file which was in the
     scope of the enclosing package class. Furthermore, because the
     owner of that symbol is the package object class that has now
     been noted as corresponding to a source file in the current
     run, the class-file backed module symbol is *also* deemed to
     be from the current run. (This logic is in `Run#compiles`.)

     Thinking the companion module already existed, no synthetic
     module was created, which would lead to a crash in extension
     methods, which needs to add methods to it.

  2. In cases when the code explicitly contains the companion pair,
     we still ran into problems in the backend whereby the class-file
     based and source-file based symbols for the module ended up in
     the same scope (of the package class). This tripped an assertion
     in `Symbol#companionModule`.

We get into these problems because of the eager manner in which
class-file based package object are opened in `openPackageModule`.
The members of the module are copied into the scope of the enclosing
package:

    scala> ScalaPackage.info.member(nme.List)
    res0: $r#59116.intp#45094.global#28436.Symbol#29451 = value List#2462

    scala> ScalaPackage.info.member(nme.PACKAGE).info.member(nme.List)
    res1: $r#59116.intp#45094.global#28436.Symbol#29451 = value List#2462

This seems to require a two-pronged defense:

  1. When we attach a pre-existing symbol for a package object symbol
     to the tree of its new source, unlink the "forwarder" symbols
     (its decls from the enclosing
     package class.

  2. In `Flatten`, in the spirit of `replaceSymbolInCurrentScope`,
     remove static member modules from the scope of the enclosing
     package object (aka `exitingFlatten(nestedModule.owner)`).

This commit also removes the warnings about defining companions
in package objects and converts those neg tests to pos (with
-Xfatal-warnings to prove they are warning free.)

Defining nested classes/objects in package objects still has
a drawback: you can't shift a class from the package to the
package object, or vice versa, in a binary compatible manner,
because of the `package$` prefix on the flattened name of
nested classes. For this reason, the `-Xlint` warning about
this remains. This issue is tracked as SI-4344.

However, if one heeds this warning and incrementatlly recompiles,
we no longer need to run into a DoubleDefinition error (which
was dressed up with a more specific diagnostic in SI-5760.)
The neg test case for that bug has been converted to a pos.

retronym committed:
SI-5954 Invalidate TypeRef cache when opening package object

I noticed that the pos/5954d was tripping a println "assertion".
This stemmed from an inconsistency between
`TypeRef#{parents, baseTypeSeq}` for a package objects compiled
from source that also has a class file from a previous compilation
run.

I've elevated the println to a devWarning, and changed
`updatePosFlags`, the home of this evil symbol overwriting,
to invalidate the caches in the symbols info. Yuck.

I believe that this symptom is peculiar to package objects because
of the way that the completer for packages calls `parents` during
the Namer phase for package objects, before we switch the symbol
to represent the package-object-from-source. But it seems prudent
to defensively invalidate the caches for any symbol that finds its
way into `updatePosFlags`.

soc commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1796,32 +1796,6 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-
-      if (mdef.symbol.isPackageObject)
-        warnPackageObjectMembers(mdef)
-
```
Nice!

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -436,6 +443,10 @@ trait Namers extends MethodSynthesis {
+        if (m.isPackageObject) {
+          val packageScope = m.enclosingPackageClass.rawInfo.decls
+          packageScope.filter(_.owner != m.enclosingPackageClass).toList.foreach(packageScope unlink _)
+        }
```
@retronym Would it be possible for us to not invoke a class-based completer for package objects in the first place? How about we remember what package objects we have in sources when traversing compilation units in namer? That would obviate the need to do something and then undo that something.

retronym replied:
I think that's the intent of @paulp's patch that you can find in #2789. Look at the first commit in particular.

I agree that's a preferable approach. It looks a touch expensive to me as currently implemented (it does a full traversal of all unit bodies for each missing symbol, as compared to `openPackageObjectsTraverser` which is just run once per-unit, albeit just _after_ namer.

xeno-by replied:
Can we cache necessary information?

retronym replied:
For one think we shouldn't use `Tree#foreach`, and rather use a custom `Traverser` that doesn't descend into `ClassDef` / `ModuleDef`. That's what `openPackageObjectsTraverser` does.

We'll need this info on every compile run (every time we open a package we look for the corresponding package class), so it seems to me that we should just move the `packageobjects` before `namer`, rather then calling it during `namer` and caching the results. But there are no doubt complications. /cc @lrytz

xeno-by replied:
Sounds good!

retronym replied:
I think that sort of thing is in 2.12 territory, btw.

xeno-by replied:
I'm a bit worried that we're going to put in a workaround that would perpetuate the current state of art, and then wouldn't find time to refactor it, as there will no longer be a need in a fix.

xeno-by replied:
Also, if talking about risks, the proposed approach of unlinking previously linked stuff that might propagate who knows where, doesn't strike me as particularly risk-free either. Maybe, if we're willing to take risks, we could take them for a cleaner solution. Or, alternatively, how about delaying this fix until 2.11.1, where it could be dealt with properly without endangering 2.11.0-final?

retronym replied:
The more radical change is still needed for SI-4695.

This one is actually pretty straight forward. It became much clearer to me when Martin explained how members of package objects are copied into the enclosing package class scope. I do welcome proposals for additional test cases, but I don't think we should delay this.

xeno-by replied:
Yes, it looks straightforward, but it adds its share of mutations to already mutation-heavy namer, which is quite scary. Maybe we could take more time to explore the alternative?

retronym replied:
I don't believe that the alternative would be likely to be merged at this juncture. This mutation is relatively easy to reason about IMO: when we `openPackageObjectTraverse` reuses the package object symbol for a source completer, remove all package-object owned symbols from the enclosing package class. It addresses an oversight of the previous implementation.

xeno-by replied:
I see. Then how about writing a TODO mentioning an alternative to explore in the future?

retronym replied:
Done

retronym committed:
SI-5954 Add a TODO comment with an idea for future work

As the prophet once said:

  "'Tis better to never do at all than to have do and undo"

Let's try that in 2.12.

retronym commented:
Review by @lrytz, perhaps?

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -436,6 +443,17 @@ trait Namers extends MethodSynthesis {
+        // This code accounts for the way the package objects found in the classpath are opened up
+        // early by the completer of the package itself. If the `packageobjects` phase then finds
+        // the same package object in sources, we have to clean the slate and remove package object
+        // members from the package class.
```
could this be done in the `packageobjects` phase to treat the same concern in one place?

retronym replied:
That seems to be late: the namer has already noticed the stale members in `ensureCompanionObject` and decided it doesn't need to create a synthetic companion.

lrytz commented on src/compiler/scala/tools/nsc/transform/Flatten.scala:
```diff
@@ -20,12 +20,18 @@ abstract class Flatten extends InfoTransform {
-    scope enter sym
+    old foreach {entry =>
+      scope unlink entry
+    }
```
this diff seems to be the same as just deleting the second line (`scope enter sym`).

retronym replied:
fixed

lrytz commented on src/compiler/scala/tools/nsc/transform/Flatten.scala:
```diff
@@ -71,9 +77,9 @@ abstract class Flatten extends InfoTransform {
-                  sym.moduleClass setFlag LIFTED
+                  decls1 enter sym
+                  if (sym.isModule)
+                    sym.moduleClass setFlag LIFTED
```
indentation

retronym replied:
fixed

lrytz commented:
The changes LGTM but i don't understand yet how they fix the bug :-)
According to the stack trace, `makeExtensionMethodSymbol` fails to enter an extension method into the scope of a companion because it is `EmptyScope`. How does the example trigger that?

```
package object pkg {
  class AnyOps(val x: Any) extends AnyVal
}
```

retronym committed:
SI-8134 Address pull request feedback

Aesthetics only.

xeno-by commented:
@retronym ping

retronym commented:
@lrytz `makeExtensionMethodSymbol` finds an empty scope because `ensureCompanionObject` decided _not_ to create a synthetic companion object for the implicit class because `companionSymbolOf` mistakenly found the forwarder symbol in the scope of the enclosing package class. This was put there in `openPackageModule`, based on _before_ we named the package object from source.

During the namer, the search for companions routes through scopes in enclosing contexts, rather then `owner.info.decls`. That's to avoid cycles. But that's why it is fooled into thinking a forwarder symbol in the package class info could be a companion object of an implicit class nested in a package object in source.

phew!

retronym commented:
ping @lrytz

lrytz commented:
LGTM.

Phew indeed. How about prohibiting class and object definitions inside package objects? Is there a ticket for that?

Sorry i didn't see the reply, I improved my gmail filters.

retronym commented:
Seeing as implicit classes can't be declared at the top level we almost encourage people to declare them in a package object.

I don't think there is anything fundamentally wrong with it, its just we were very light on testing for separate compilation.

https://issues.scala-lang.org/browse/SI-4344 tracks a desirable change, which is to give package-object nested classes/objects the same post-flatten names as eponymous top-level classes. This would mean you could move them in and out of the package object without breaking binary compatibility.

lrytz commented:
> Seeing as implicit classes can't be declared at the top level we almost encourage people to declare them in a package object.

OK. Maybe we could lift that restriction by putting the generated implicit def into a (synthetic) (package) object.

retronym commented:
Separate compilation makes that pretty hard.

It might have been cleaner to spec that the constructor of the implicit class is itself implicit, rather than desugaring to an implicit class+def. The desugaring was already the source of interesting bugs for things like implicit case- or value-classes, where you end up with a factory method and a companion with the same name. And if the companion is nested, it might later need a module accessor method.
